File: backend/src/github/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes, createHmac, timingSafeEqual } from "node:crypto";
import { githubConfig } from "./config.js";

const b64u = (b: Buffer) => b.toString("base64url");
const fromB64u = (s: string) => Buffer.from(s, "base64url");

export function encryptToken(plain: string): string {
  const { encKey } = githubConfig();
  const iv = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", encKey, iv);
  const data = Buffer.concat([cipher.update(plain, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return [b64u(iv), b64u(tag), b64u(data)].join(".");
}

export function decryptToken(payload: string): string {
  const { encKey } = githubConfig();
  const [iv, tag, data] = payload.split(".").map(fromB64u);
  const decipher = createDecipheriv("aes-256-gcm", encKey, iv);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(data), decipher.final()]).toString("utf8");
}

function hmac(input: string): string {
  return b64u(createHmac("sha256", githubConfig().stateSecret).update(input).digest());
}

export function signState(ttlMs = 10 * 60 * 1000): string {
  const payload = b64u(Buffer.from(JSON.stringify({ nonce: b64u(randomBytes(8)), exp: Date.now() + ttlMs })));
  return `${payload}.${hmac(payload)}`;
}

export function verifyState(state: string): boolean {
  const [payload, sig] = state.split(".");
  if (!payload || !sig) return false;
  const expected = hmac(payload);
  if (sig.length !== expected.length || !timingSafeEqual(Buffer.from(sig), Buffer.from(expected))) return false;
  try {
    const { exp } = JSON.parse(fromB64u(payload).toString("utf8"));
    return typeof exp === "number" && Date.now() < exp;
  } catch { return false; }
}

File: backend-py/app/github/crypto.py
import base64
import hmac
import json
import os
import time
from hashlib import sha256

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .config import github_config


def _b64u(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).rstrip(b"=").decode()


def _from_b64u(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def encrypt_token(plain: str) -> str:
    key = github_config().enc_key
    iv = os.urandom(12)
    ct = AESGCM(key).encrypt(iv, plain.encode(), None)  # ciphertext || 16-byte tag
    data, tag = ct[:-16], ct[-16:]
    return ".".join([_b64u(iv), _b64u(tag), _b64u(data)])


def decrypt_token(payload: str) -> str:
    key = github_config().enc_key
    iv_s, tag_s, data_s = payload.split(".")
    iv, tag, data = _from_b64u(iv_s), _from_b64u(tag_s), _from_b64u(data_s)
    return AESGCM(key).decrypt(iv, data + tag, None).decode()


def _hmac(payload: str) -> str:
    secret = github_config().state_secret.encode()
    return _b64u(hmac.new(secret, payload.encode(), sha256).digest())


def sign_state(ttl_ms: int = 10 * 60 * 1000) -> str:
    body = {"nonce": _b64u(os.urandom(8)), "exp": int(time.time() * 1000) + ttl_ms}
    payload = _b64u(json.dumps(body).encode())
    return f"{payload}.{_hmac(payload)}"


def verify_state(state: str) -> bool:
    parts = state.split(".")
    if len(parts) != 2 or not parts[0] or not parts[1]:
        return False
    payload, sig = parts
    expected = _hmac(payload)
    if not hmac.compare_digest(sig, expected):
        return False
    try:
        body = json.loads(_from_b64u(payload).decode())
        return isinstance(body.get("exp"), int) and time.time() * 1000 < body["exp"]
    except Exception:
        return False
